[FLINK-22177][table-api/doc] Add docs for time functions and time zone support (#15634)
apache/flink: docs/content/docs/dev/table/concepts/timezone.md

---

> title: "Timestamp DataTypes and Time Zone Support"

Author: I learned this from other vendors, I'm ok to rename to `Time Zone`

+----+----------------------+-------------+-------------------------+-------------------------+
```

```sql
Flink SQL> SELECT c, CAST(c AS TIMESTAMP(3)), CAST(c as STRING), CAST(d AS TIMESTAMP_LTZ(3)) FROM T1;
```

```
+----+-------------------------+-------------------------+-------------------------+-------------------------+
| op |                       c |                  EXPR$1 |               EXPR$2    |                  EXPR$3 |
+----+-------------------------+-------------------------+-------------------------+-------------------------+
| +I | 1970-01-01 08:00:04.001 | 1970-01-01 08:00:04.001 | 1970-01-01 08:00:04.001 | 1970-01-01 00:00:01.001 |
+----+-------------------------+-------------------------+-------------------------+-------------------------+
```

## Time Attribute and Time Zone

Please see [Time Attribute]({{< ref "docs/dev/table/concepts/time_attributes">}}#date-and-time) for more information about time attribute.

### Processing Time Attribute and Time Zone
Flink SQL definine process time attribute by function PROCTIME(), the function return type is TIMESTAMP_LTZ.

> **NOTE**
> Before Flink 1.13, the function return type of PROCTIME() is TIMESTAMP, and the return value is the TIMESTAMP in UTC time zone,
> E.g. the PROCTIME() returns `2021-03-01 04:00:00` when user is in Shanghai and current timestamp is `2021-03-01 12:00:00 (UTC+8)`. 
> Flin 1.13 fixes this issue and uses TIMESTAMP_LTZ type as return type of PROCTIME(), users don't need to deal time zone proctime anymore.

Reviewer: ```suggestion
{{< hint info >}}
Before Flink 1.13, the function return type of `PROCTIME()` is `TIMESTAMP`, and the return value is the `TIMESTAMP` in UTC time zone,
e.g. the wall-clock shows `2021-03-01 12:00:00` at Shanghai, however the `PROCTIME()` displays `2021-03-01 04:00:00` which is wrong.
Flin 1.13 fixes this issue and uses `TIMESTAMP_LTZ` type as return type of `PROCTIME()`, users don't need to deal time zone problems anymore.
{{< /hint >}}
```

The PROCTIME() always represents your local timestamp value, using TIMESTAMP_LTZ type can also support DayLight Saving Time well. 

```sql
Flink SQL> SET table.local-time-zone=UTC; 
Flink SQL> SELECT PROCTIME();
```
```
+----+-------------------------+
| op |                  EXPR$0 |
+----+-------------------------+
| +I | 2021-04-15 14:48:31.387 |
+----+-------------------------+
```

```sql
Flink SQL> SET table.local-time-zone=UTC+8; 
Flink SQL> SELECT PROCTIME();
```
```
+----+-------------------------+
| op |                  EXPR$0 |
+----+-------------------------+
| +I | 2021-04-15 22:48:34.567 |
+----+-------------------------+
```
```sql
Flink SQL> CREATE TABLE src (
             a INT,
             b DOUBLE,
             proctime as PROCTIME()
           ) WITH (
             'connector' = 'kafka',
             ...);
Flink SQL>  SELECT 
                TUMBLE_START(proctime, INTERVAL '0.003' SECOND),
                TUMBLE_END(proctime, INTERVAL '0.003' SECOND),
                TUMBLE_PROCTIME(proctime, INTERVAL '0.003' SECOND),
                COUNT(a),
                SUM(b)
            FROM src

>             GROUP BY TUMBLE(proctime, INTERVAL '0.003' SECOND);

Reviewer: Better to use `10` minutes as window interval, which is easier for example.

 -- the TUMBLE_START    column will be close to 2021-04-15 22:48:34.567, the type of TUMBLE_START    column is TIMESTAMP
 -- the TUMBLE_END      column will be close to 2021-04-15 22:48:34.570, the type of TUMBLE_END      column is TIMESTAMP      

>  -- the TUMBLE_PROCTIME column will be close to 2021-04-15 22:48:34.570, the type of TUMBLE_PROCTIME column is TIMESTAMP_LTZ      

Reviewer: This is hard to understand. Would be better to show some result example here and explain return type of `TUMBLE_START`, `TUMBLE_END`, `TUMBLE_PROCTIME` separately.

          
```

### Event Time Attribute and Time Zone
Flink supports defining event time attribute on TIMESTAMP column and TIMESTAMP_LTZ column. It's recommended to defining event time attribute on TIMESTAMP column if the data source contains TIMESTAMP literal.  
```sql
Flink SQL> CREATE TABLE src (
             ts TIMESTAMP(3), -- TIMESTAMP DataType
             a INT,
             b DOUBLE,
             WATERMARK FOR ts AS ts - INTERVAL '0.001' SECOND
           ) WITH (
             'connector' = 'kafka',
                          ...);
Flink SQL>  SELECT
                TUMBLE_START(ts, INTERVAL '0.003' SECOND), 
                TUMBLE_END(ts, INTERVAL '0.003' SECOND),
                TUMBLE_ROWTIME(ts, INTERVAL '0.003' SECOND),
                COUNT(a),
                SUM(b)
            FROM src
            GROUP BY TUMBLE(ts, INTERVAL '0.003' SECOND); 

>  -- the TUMBLE_START    column will be close to ts,                           the type of TUMBLE_START    column is TIMESTAMP

Reviewer: ditto.

             ts BIGINT, -- The epoch time in milliseconds, BIGINT DataType
             a INT,
             b DOUBLE,
             ts_ltz AS TO_TIMESTAMP_LTZ(ts, 3), -- to TIMESTAMP_LTZ DataType
             WATERMARK FOR ts_ltz AS ts_ltz - INTERVAL '0.001' SECOND
           ) WITH (
             'connector' = 'kafka',
                         ...);
Flink SQL>  SELECT
                TUMBLE_START(ts_ltz, INTERVAL '0.003' SECOND), 
                TUMBLE_END(ts_ltz, INTERVAL '0.003' SECOND),
                TUMBLE_ROWTIME(ts_ltz, INTERVAL '0.003' SECOND),
                COUNT(a),
                SUM(b)            FROM src
            GROUP BY TUMBLE(ts_ltz, INTERVAL '0.003' SECOND);
 -- the TUMBLE_START    column will be close to ts_ltz (represents timestamp value in Asia/Shanghai), the type of TUMBLE_START    column is TIMESTAMP
 -- the TUMBLE_END      column will be close to ts_ltz + INTERVAL '0.003' SECOND,             the type of TUMBLE_END      column is TIMESTAMP      
 -- the TUMBLE_ROWTIME  column will be close to ts_ltz + INTERVAL '0.002' SECOND,             the type of TUMBLE_ROWTIME  column is TIMESTAMP_LTZ   
            
```

## Daylight Saving Time Support

> Flink SQL supports defining time attributes on TIMESTAMP_LTZ column, base on this, Flink SQL gently uses TIMESTAMP and TIMESTAMP_LTZ type in window processing to support the Daylight Saving Time.

Reviewer: gently => gracefully?

> Flink use timestamp literal to split the window and assigns window to data according to the epoch time of the each row. It means Flink uses TIMESTAMP type for window start and window end(e.g. TUMBLE_START and TUMBLE_END), uses TIMESTAMP_LTZ for window time attribute(e.g. TUMBLE_PROCTIME, TUMBLE_ROWTIME).

Reviewer: ```suggestion
Flink use timestamp literal to split the window and assigns window to data according to the epoch time of the each row. It means Flink uses `TIMESTAMP` type for window start and window end (e.g. `TUMBLE_START` and `TUMBLE_END`), uses `TIMESTAMP_LTZ` for window time attribute (e.g. `TUMBLE_PROCTIME`, `TUMBLE_ROWTIME`).
```

> Giving a example of tumble window, The DaylightTime in Los_Angele start at time 2021-03-14 02:00:00:

Reviewer: ```suggestion
Given a example of tumble window, the DaylightTime in Los_Angele starts at time 2021-03-14 02:00:00:
```

```
long epoch1 = 1615708800000L; // 2021-03-14 00:00:00
long epoch2 = 1615712400000L; // 2021-03-14 01:00:00
long epoch3 = 1615716000000L; // 2021-03-14 03:00:00, skip one hour (2021-03-14 02:00:00)
long epoch4 = 1615719600000L; // 2021-03-14 04:00:00 
```
The tumble window [2021-03-14 00:00:00,  2021-03-14 00:04:00] will collect 3 hours' data in Los_angele time zone, but it collect 4 hours' data in other non-DST time zones, what user to do is only define time attribute on TIMESTAMP_LTZ column.

All windows in Flink like Hop window, Session window, Cumulative window follow this way, and all operations in Flink SQL support TIMESTAMP_LTZ well, thus Flink gracefully supports the Daylight Saving Time zone.   

## Difference between Batch and Streaming Mode
The following time functions:
* LOCALTIME
* LOCALTIMESTAMP
* CURRENT_DATE
* CURRENT_TIME
* CURRENT_TIMESTAMP
* NOW()

> Flink evaluates their values according to execution mode, i.e. Flink evaluates time function value for row level in Streaming mode, evaluates the time function value at query start for batch mode.

Reviewer: ```suggestion
Flink evaluates their values according to execution mode. They are evaluated for each record in streaming mode. But in batch mode, they are evaluated once as the query starts and uses the same result for every row.
```
